tests/handler/handler_copy.cpp

```diff
@@ -17,334 +22,886 @@ using mode_t = cl::sycl::access::mode;
+          cgh.copy(ctx.getSrcHostPtr(), w);
+        },
+        lh.set_line(__LINE__).set_op(
+            "copy(shared_ptr_class<$dataT>, accessor<$dataT, $dim_dst, "
+            "$mode_dst, $target>)"));
   }
   {
     // Check copy(dataT*, accessor)
-    const auto func = [](cl::sycl::handler& cgh, acc_read_t accRead,
-                         acc_write_t accWrite, host_shared_ptr hostPtrRead,
-                         host_shared_ptr hostPtrWrite) {
-      cgh.copy(hostPtrRead.get(), accWrite);
-    };
-    TEST_SINGLE_COPY_FUNCTION_ACC_WRITE(dataT, dim_src, dim_dst, readMode,
-                                        writeMode, target, queue, func);
+    copy_test_context<dataT, dim_src, dim_dst, strided, transposed> ctx(queue);
+    ctx.verify_h2d_copy(
+        [&](cl::sycl::handler& cgh) {
+          auto w = ctx.getDstBuf().template get_access<mode_dst, target>(
+              cgh, ctx.getDstCopyRange(), ctx.getDstCopyOffset());
+          cgh.copy(ctx.getSrcHostPtr().get(), w);
+        },
+        lh.set_line(__LINE__).set_op(
+            "copy($dataT*, accessor<$dataT, $dim_dst, $mode_dst, $target>)"));
   }
   {
     // Check copy(accessor, accessor)
-    const auto func = [](cl::sycl::handler& cgh, acc_read_t accRead,
-                         acc_write_t accWrite, host_shared_ptr hostPtrRead,
-                         host_shared_ptr hostPtrWrite) {
-      cgh.copy(accRead, accWrite);
-    };
-    TEST_SINGLE_COPY_FUNCTION_ACC_WRITE(dataT, dim_src, dim_dst, readMode,
-                                        writeMode, target, queue, func);
+    copy_test_context<dataT, dim_src, dim_dst, strided, transposed> ctx(queue);
+    ctx.verify_d2d_copy(
+        [&](cl::sycl::handler& cgh) {
+          auto r = ctx.getSrcBuf().template get_access<mode_src, target>(
+              cgh, ctx.getSrcCopyRange(), ctx.getSrcCopyOffset());
+          auto w = ctx.getDstBuf().template get_access<mode_dst, target>(
+              cgh, ctx.getDstCopyRange(), ctx.getDstCopyOffset());
+          cgh.copy(r, w);
+        },
+        lh.set_line(__LINE__).set_op(
+            "copy(accessor<$dataT, $dim_src, $mode_src, $target>, "
+            "accessor<$dataT, $dim_dst, $mode_dst, $target>)"));
   }
   {
     // Check fill(accessor, dataT)
-    const auto func = [](cl::sycl::handler& cgh, acc_read_t accRead,
-                         acc_write_t accWrite, host_shared_ptr hostPtrRead,
-                         host_shared_ptr hostPtrWrite) {
-      const auto pattern = dataT(117);
-      cgh.fill(accWrite, pattern);
-    };
-    TEST_SINGLE_COPY_FUNCTION_ACC_WRITE(dataT, dim_src, dim_dst, readMode,
-                                        writeMode, target, queue, func);
+    const auto pattern = type_helper<dataT>::make(117);
+    copy_test_context<dataT, dim_src, dim_dst, strided, transposed> ctx(queue);
+    ctx.verify_fill(
+        [&](cl::sycl::handler& cgh) {
+          auto w = ctx.getDstBuf().template get_access<mode_dst, target>(cgh);
```

Comment:
Seems like there's a bug here: the accessor is created without range or offset, but `verify_fill` checks the results as if it was a ranged accessor.

Reply:
Oof, you are right! That might offer an explanation for why all implementations wrote the full range...